keps/sig-apps/19-Graduate-CronJob-to-Stable/README.md

```diff
@@ -0,0 +1,605 @@
+    - [Cron Aspect](#cron-aspect)
+- [Improvements/Considerations](#improvementsconsiderations)
+  - [Add .status.nextScheduleTime](#add-statusnextscheduletime)
+  - [Add counters](#add-counters)
+  - [Add .status.conditions](#add-statusconditions)
+  - [Support Jitter for cronjobs](#support-jitter-for-cronjobs)
+  - [Support Timezone for cronjobs](#support-timezone-for-cronjobs)
+  - [CronJob umbrella issue](#cronjob-umbrella-issue)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [x] "Implementation History" section is up-to-date for milestone
+- [x] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [x] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+[CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) is a Kubernetes API that creates Job object on a schedule specified by a cron spec. It is in beta status since v1.8 and currently it is marked deprecated with target release for removal being v1.22. This document lays out the plan to promote it to stable.
+
+## Motivation
+
+CronJob is useful to run periodic tasks using cron like facility in a kubernetes cluster. Its `.spec` has been stable for the last few releases. We feel the API with additional `.status` information is ready to be promoted to Stable and be supported long-term by the community.
+
+### Goals
+
+- Write a new controller which will
+  - Use informers instead of polling
+  - Address some open issues with the current controller
+  - Add metrics exposing controller throughput, latency etc.
+- Extend CronJob status field
+  - lastSuccessfulTime: tracks the last time the job completed successfully
+
+## Proposal
+
+### Plan for promoting CronJob to GA
+
+To promote to GA we would create `batch/v1/CronJob` in the [batch/v1 API](https://github.com/kubernetes/kubernetes/blob/b58777eda0f0077eba79d063c1c8f31cd7dbb8b9/staging/src/k8s.io/api/batch/v1/types.go).
+
+We shall have dual implementation of controller (old and new) co-exist, until the new controller reaches GA.
+The old and new implementation of the controller will be toggled through a feature flag:
+- alpha (disabled by default, can be enabled)
+- beta (enabled by default, can be disabled)
+- GA (enabled by default, cannot be disabled)
```

Comment:
nit: maybe make explicit that we will remove the old implementation at this point (or soon after)

Reply:
Done.